src/main/java/com/datadoghq/sketch/ddsketch/store/PaginatedStore.java

```diff
@@ -0,0 +1,387 @@
+                        }
+                    }
+                }
+            }
+        }
+        throw new NoSuchElementException();
+    }
+
+    @Override
+    public double getTotalCount() {
+        if (isEmpty()) {
+            return 0D;
+        }
+        double total = 0D;
+        for (double[] page : pages) {
+            if (null != page) {
+                for (double count : page) {
+                    total += count;
+                }
+            }
+        }
+        return total;
+    }
+
+    @Override
+    public void add(int index, double count) {
+        if (count > 0) {
+            int alignedIndex = alignedIndex(index);
+            double[] page = getPage(alignedIndex >>> PAGE_SHIFT);
+            page[alignedIndex & PAGE_MASK] += count;
+        }
+    }
+
+    private double[] getPage(int pageIndex) {
+        double[] page = pages[pageIndex];
+        if (null == page) {
+            page = pages[pageIndex] = new double[PAGE_SIZE];
+        }
+        return page;
+    }
+
+    private int alignedIndex(int index) {
+        // get the index of the page this value should be stored in
+        int pageIndex = index < 0
+                ? ~(-index >>> PAGE_SHIFT)
+                : index >>> PAGE_SHIFT;
+        if (pageIndex < minPageIndex) {
+            // then space needs to be made before the first page,
+            // unless this is the first insertion
+            if (null == pages) {
+                lazyInit(pageIndex);
+            } else {
+                shiftPagesRight(pageIndex);
+            }
+        } else if (pageIndex >= minPageIndex + pages.length - 1) {
+            // then space needs to be made after the last page
+            extendTo(pageIndex);
+        }
+        // align the index relative to the start of the sketch
+        return index + (-minPageIndex << PAGE_SHIFT);
+    }
+
+    private void lazyInit(int pageIndex) {
+        minPageIndex = pageIndex;
+        pages = new double[GROWTH][];
+    }
+
+    private void shiftPagesRight(int pageIndex) {
+        int requiredExtension = minPageIndex - pageIndex;
+        // check if there is space to shift into
+        boolean canShiftRight = true;
+        for (int i = 0; i < requiredExtension && canShiftRight; ++i) {
+            canShiftRight = null == pages[pages.length - i - 1];
+        }
+        if (canShiftRight) {
+            System.arraycopy(pages, 0, pages, requiredExtension, pages.length - requiredExtension);
+        } else {
+            double[][] newPages = new double[pages.length + aligned(requiredExtension)][];
+            System.arraycopy(pages, 0, newPages, requiredExtension, pages.length);
+            this.pages = newPages;
+        }
+        Arrays.fill(pages, 0, requiredExtension, null);
+        this.minPageIndex = pageIndex;
+    }
+
+    private void extendTo(int pageIndex) {
+        this.pages = Arrays.copyOf(pages, pages.length + aligned(pageIndex - minPageIndex + 1));
```

Comment:
I'm not certain about the size going into aligned. Shouldn't it take the `pages.length` into account as well?

Reply:
`pages` always has an aligned length.

Reply:
Yes you're right, and fixing this saves a little bit of space too